src/Stratis.FederatedPeg.Features.FederationGateway/Wallet/FederationWalletManager.cs

```diff
@@ -993,35 +980,65 @@ private void UpdateWhenChainDownloaded(IEnumerable<FederationWallet> wallets, Da
         }
 
         /// <inheritdoc />
-        public void ImportMemberKey(string password, string mnemonic, string passphrase)
+        public void EnableFederation(string password, string mnemonic = null, string passphrase = null, bool onlyImportKey = false)
         {
             Guard.NotEmpty(password, nameof(password));
-            Guard.NotEmpty(mnemonic, nameof(mnemonic));
 
-            // Get the extended key.
-            ExtKey extendedKey;
-            try
+            // Protect against de-activation if the federation is already active.
+            if (this.isFederationActive)
             {
-                extendedKey = HdOperations.GetExtendedKey(mnemonic, passphrase);
+                this.logger.LogWarning("(-):[FEDERATION_ALREADY_ACTIVE]");
+                return;
             }
-            catch (NotSupportedException ex)
+
+            // Get the extended key.
+            if (!string.IsNullOrEmpty(mnemonic))
             {
-                this.logger.LogTrace("Exception occurred: {0}", ex.ToString());
-                this.logger.LogTrace("(-)[EXCEPTION]");
+                ExtKey extendedKey;
+                try
+                {
+                    extendedKey = HdOperations.GetExtendedKey(mnemonic, passphrase);
+                }
+                catch (NotSupportedException ex)
+                {
+                    this.logger.LogTrace("Exception occurred: {0}", ex.ToString());
+                    this.logger.LogTrace("(-)[EXCEPTION]");
+
+                    if (ex.Message == "Unknown")
+                        throw new WalletException("Please make sure you enter valid mnemonic words.");
 
-                if (ex.Message == "Unknown")
-                    throw new WalletException("Please make sure you enter valid mnemonic words.");
+                    throw;
+                }
 
-                throw;
+                // Create a wallet file.
+                string encryptedSeed = extendedKey.PrivateKey.GetEncryptedBitcoinSecret(password, this.network).ToWif();
+
+                this.Wallet.EncryptedSeed = encryptedSeed;
+                this.SaveWallet();
             }
 
-            // Create a wallet file.
-            string encryptedSeed = extendedKey.PrivateKey.GetEncryptedBitcoinSecret(password, this.network).ToWif();
+            if (onlyImportKey)
+                return;
 
-            this.Wallet.EncryptedSeed = encryptedSeed;
-            this.SaveWallet();
+            try
+            {
+                Key key = Key.Parse(this.Wallet.EncryptedSeed, password, this.Wallet.Network);
 
-            this.logger.LogTrace("(-)");
+                this.Secret = new WalletSecret() { WalletPassword = password };
+                this.isFederationActive = key.PubKey.ToHex() == this.federationGatewaySettings.PublicKey;
```

Comment:
Can we check that this is true before we save the encrypted seed? `key.PubKey.ToHex() == this.federationGatewaySettings.PublicKey;`

Otherwise we'll write the encrypted seed but then we won't be able to overwrite it.

Reply:
Done